psyneulink/core/components/mechanisms/modulatory/control/optimizationcontrolmechanism.py

```diff
@@ -1877,9 +1877,18 @@ def _execute(self, variable=None, context=None, runtime_params=None):
 
         # freeze the values of current context, because they can be changed in between simulations,
         # and the simulations must start from the exact spot
-        self.agent_rep._initialize_from_context(self._get_frozen_context(context),
-                                                base_context=context,
-                                                override=True)
+        frozen_context = self._get_frozen_context(context)
+
+        alt_controller = None
+        if self.agent_rep.controller is None:
+            try:
+                alt_controller = context.composition.controller
+            except AttributeError:
```

Comment:
is this supposed to catch both missing `context.composition` and `context.composition.controller`?
if it's only the latter.
`alt_controller = getattr(context.composition, 'controller', None)`
would be more readable.

Reply:
It was meant to catch both. I'm unsure if there's a case where this code would be run in a context with no composition set, but if that happened I think it should handle it.

Reply:
is the absence of `composition` in `context` legal in this situation? or is it punting the error to be reported later?

Reply:
I think in that case, it's ok and intended not to deal with it. alt_controller would just be one other possibility to try in the case that agent_rep has no controller. If it's not ok for context to have no composition, then I think that should be reported elsewhere.